lib/charms/tls_certificates_interface/v4/tls_certificates.py

```diff
@@ -0,0 +1,1211 @@
+        )
+        locality_name = certificate_object.subject.get_attributes_for_oid(NameOID.LOCALITY_NAME)
+        organization_name = certificate_object.subject.get_attributes_for_oid(
+            NameOID.ORGANIZATION_NAME
+        )
+        email_address = certificate_object.subject.get_attributes_for_oid(NameOID.EMAIL_ADDRESS)
+        try:
+            sans = certificate_object.extensions.get_extension_for_class(
+                x509.SubjectAlternativeName
+            ).value
+            sans_dns = [
+                str(san)
+                for san in sans.get_values_for_type(x509.DNSName)
+                if isinstance(san, x509.DNSName)
+            ]
+            sans_ip = [
+                str(san)
+                for san in sans.get_values_for_type(x509.IPAddress)
+                if isinstance(san, x509.IPAddress)
+            ]
+            sans_oid = [
+                str(san)
+                for san in sans.get_values_for_type(x509.RegisteredID)
+                if isinstance(san, x509.RegisteredID)
+            ]
+        except x509.ExtensionNotFound:
+            sans_dns = []
+            sans_ip = []
+            sans_oid = []
+        expiry_time = certificate_object.not_valid_after_utc
+        validity_start_time = certificate_object.not_valid_before_utc
+
+        return Certificate(
+            raw=certificate,
+            common_name=str(common_name[0].value),
+            country_name=str(country_name[0].value) if country_name else None,
+            state_or_province_name=str(state_or_province_name[0].value)
+            if state_or_province_name
+            else None,
+            locality_name=str(locality_name[0].value) if locality_name else None,
+            organization=str(organization_name[0].value) if organization_name else None,
+            email_address=str(email_address[0].value) if email_address else None,
+            sans_dns=sans_dns,
+            sans_ip=sans_ip if sans_ip else None,
+            sans_oid=sans_oid if sans_oid else None,
+            expiry_time=expiry_time,
+            validity_start_time=validity_start_time,
+        )
+
+
+@dataclass
+class ProviderCertificate:
+    """This class represents a certificate provided by the TLS provider."""
+
+    certificate: Certificate
+    certificate_signing_request: CertificateSigningRequest
+    ca: Certificate
+    chain: List[Certificate]
+    recommended_expiry_notification_time: Optional[int] = None
+    revoked: Optional[bool] = None
+
+
+class CertificateAvailableEvent(EventBase):
+    """Charm Event triggered when a TLS certificate is available."""
+
+    def __init__(
+        self,
+        handle: Handle,
+        certificate: str,
+        certificate_signing_request: str,
+        ca: str,
+        chain: List[str],
+    ):
+        super().__init__(handle)
+        self.certificate = certificate
+        self.certificate_signing_request = certificate_signing_request
+        self.ca = ca
+        self.chain = chain
```

Comment:
Should the event have the object versions (or a `ProviderCertificate`), since it is a public interface?

Reply:
Events must only contain simple types, we can't have something like:

```
class CertificateAvailableEvent(EventBase):
    """Charm Event triggered when a TLS certificate is available."""

    def __init__(
        self,
        handle: Handle,
        certificate: Certificate,
        certificate_signing_request: CertificateSigningRequest,
        ca: Certificate,
        chain: List[Certificate],
    ):
        super().__init__(handle)
        self.certificate = certificate
        self.certificate_signing_request = certificate_signing_request
        self.ca = ca
        self.chain = chain

```

Reply:
I didn't know that...

Reply:
is that true in the case of custom events? I've definitely used relation before, and that isn't a simple type.

as long as the snapshot functions are implemented, I can't see why there would be an issue.

Anywhere I can read more about this?

Reply:
I was getting this error with scenario: `scenario.runtime.UncaughtCharmError: Uncaught exception (<class 'ValueError'>) in operator/charm code: ValueError("unable to save the data for CertificateAvailableEvent, it must contain only simple types: `. I didn't troubleshoot it much though, it's possible that I need to change the snapshot/restore functions.

Reply:
Ahh, ok. It looks like it is validating that whatever is returned from `snapshot()` only contains simple types.

So, we'd probably need to implement something like this:

```
    def snapshot(self) -> dict:
        """Return snapshot."""
        return {
            "certificate": str(self.certificate),
            "certificate_signing_request": str(self.certificate_signing_request),
            "ca": str(self.ca),
            "chain": json.dumps([str(certificate) for certificate in self.chain]),
        }

    def restore(self, snapshot: dict):
        """Restore snapshot."""
        self.certificate = Certificate.from_string(snapshot["certificate"])
        self.certificate_signing_request = CertificateSigningRequest.from_string(
            snapshot["certificate_signing_request"]
        )
        self.ca = Certificate.from_string(snapshot["ca"])
        chain_strs = json.loads(snapshot["chain"])
        self.chain = [Certificate.from_string(chain_str) for chain_str in chain_strs]
```

if we wanted to go that route. There's probably better ways to serialize / deserialize it, but this was quick.

Reply:
Ok I just edited the event to do just that.